external/js/cothority/src/byzcoin/byzcoin-rpc.ts

```diff
@@ -168,6 +188,70 @@ export default class ByzCoinRPC implements ICounterUpdater {
         return header.version;
     }
 
+    /**
+     * Returns an observable proof for an instance that will be
+     * automatically updated whenever the instance is changed.
+     * It returns a BehaviorSubject, which is a special observable that
+     * keeps the current value.
+     * It also caches all requests, so if there are two requests for the
+     * same id, no new observable will be created.
+     * It connects to `getNewBlocks` to be informed whenever a new block is
+     * created.
+     * @param id of the instance to return
+     * @throws an error if the instance does not exist
+     */
+    async proofObservable(id: InstanceID): Promise<BehaviorSubject<Proof>> {
+        const bs = this.cache.get(id);
+        if (bs !== undefined) {
+            return bs;
+        }
+
+        // Check if the db already has a version, which might be outdated,
+        // but still better than to wait for the network.
+        // might be old, but be informed as soon as the correct values arrive.
+        // This makes it possible to have a quick display of values that
+        const idStr = id.toString("hex");
+        const proofBuf = await this.db.get(idStr);
+        let dbProof: Proof;
+        if (proofBuf === undefined) {
+            dbProof = await this.getProofFromLatest(id);
+        } else {
+            dbProof = Proof.decode(proofBuf);
+        }
+        if (!dbProof.exists(id)) {
+            throw new Error("this instance does not exist");
+        }
+
+        // Create a new BehaviorSubject with the proof, which might not be
+        // current, but a best guess from the db of a previous session.
+        const bsNew = new BehaviorSubject(dbProof);
+        this.cache.set(id, bsNew);
+
+        // Set up a pipe from the block to fetch new versions if a new block
+        // arrives.
+        // Start with an observable that emits each new block as it arrives.
+        (await this.getNewBlocks())
+            .pipe(
+                // Make sure only newer blocks than the proof are taken into
+                // account
+                filter((block) => block.index > dbProof.latest.index),
+                // Get a new proof of the instance
+                mergeMap(() => this.getProofFromLatest(id)),
```

Comment:
Instead of calling the `getProofFromLatest`, why not checking directly if the block contains an instruction about the instanceID we observe?

Reply:
Yes, I thought about that, but it will not work everytime. If you do an `Instruction` to a `Coin` with the command `transfer`, two instances will change: the one in `InstanceID`, and the one in the argument `destination`. So it would be quite difficult to know which change.

See also #2216